### PR TITLE
CHE-4098: separate terminal and exec agents

### LINKED ISSUE #3253: Config che server to Nginx,but terminal is die
Hello, I use Nignx reverse proxy server to run my **che server** and **che workspace** link, so that the browser's entrance to look the same, and everything is running very well. 

But I open a terminal in che server, a few minutes or more without operations, terminal die(The terminal websocket broken links ), unable to input and output. can only be close, to open a new one  or  refresh the page . 

I guess whether the nginx websocket long connection timeout, no operation off within a few minutes? 

If it is for this reason is there any way to avoid? In addition to configure nginx longer timeout. Or maybe something else? 

thanks.

### LINKED ISSUE #4098: Separate exec and terminal agents
Separate exec and terminal agents.

### LINKED ISSUE #4159: Exec Agent stop responding after http access
Connect with http on /pty url of exec agent and it dies.

**Reproduction Steps:**

Start a workspace
Wait the boot of exec-agent

it displays
```
[STDOUT] Channel Routes:
[STDOUT] ✓ Connect to Exec-Agent(webscoket) ........ GET /connect
[STDOUT] 
[STDOUT] Terminal routes:
[STDOUT] ✓ Connect to pty(webscoket) ............... GET /pty
```

Connect with curl in http to the pty
```
curl http://localhost:32786/pty
curl: (52) Empty reply from server
```
(Yeah we should use websocket)
but remotely the exec agent stops with the error and exit
```
2017/02/16 15:37:38 Websocket upgrade failed: websocket: version != 13
```

**OS and version:**    
macOs sierra
Che 5.3.0

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?
Separate terminal and exec agents.
Refactor golang terminal code.
Add ping frames into terminal websocket connection.

### What issues does this PR fix or reference?
Fixes #4098 
Fixes #4159 
Fixes #3253 

#### Changelog
Separated terminal and exec agents. 
Fix terminal disconnects due to request timeout behind a proxy.
Fix terminal failure after regular HTTP request to a websocket entrypoint.

#### Release Notes
We have separated the terminal and exec agents in this release to make behaviors clearer and give users the freedom to control whether terminal, command execution or both are required for their machines.

In order to provide a smooth transition period the REST API services will automatically add the exec agent into every machine that has a terminal agent for the next two releases. This temporary service will be removed in 5.8.0. At that point any machines that do not have an exec agent added will be unable to have commands executed against them.

You can add the terminal and/or exec agent to your machine through the user dashboard in the Workspace detail view, or via the workspace API.

We have also fixed a bug in the terminal that would cause the terminal to become disconnected when a request timed out behind a proxy. Terminal and exec agents now send ping frames to clients to prevent this disconnection issue.

### CODE REVIEW

File: dockerfiles/lib/src/internal/dir/che-dir.ts [hunk 1]
```diff
@@ -846,7 +846,7 @@ setupSSHKeys(workspaceDto: org.eclipse.che.api.workspace.shared.dto.WorkspaceDto
     let promises : Array<Promise<any>> = new Array<Promise<any>>();
     let workspaceCommands : Array<any> = workspaceDto.getConfig().getCommands();
 
-    // get terminal URI
+    // get exec-agent URI
     let execAgentServer = workspaceDto.getRuntime().getDevMachine().getRuntime().getServers().get("4411/tcp");
```
Comment (AndrienkoAleksandr) [inline dockerfiles/lib/src/internal/dir/che-dir.ts]: Maybe port should be 4412?

Comment (garagatyi) [reply dockerfiles/lib/src/internal/dir/che-dir.ts]: Yep, good catch!

Comment (garagatyi) [reply dockerfiles/lib/src/internal/dir/che-dir.ts]: Fixed

**Comment by garagatyi** [comment]
@bmicklea I added release notes. Feel free to change them as you want.

**Review by skabashnyuk — APPROVED** [review]

File: agents/go-agents/src/main/go/core/rest/router.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+	"github.com/julienschmidt/httprouter"
+)
+
+// Router provides http requests routing capabilities
+type Router interface {
+	http.Handler
+}
+
+type routerImpl struct {
```
Comment (voievodin) [inline agents/go-agents/src/main/go/core/rest/router.go]: I think it makes more sense to create a separate go file and move `httprouter` based implementation there + add routes registration to the contract of the `Router` interface.

File: agents/go-agents/src/main/go/core/httputils/httputils.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+package httputils
+
+import (
+	"net/http"
+	"regexp"
+)
+
+// BasePathChopper cuts base path of a request
+type BasePathChopper struct {
```
Comment (voievodin) [inline agents/go-agents/src/main/go/core/httputils/httputils.go]: This is kind of workaround for the `httprouter` implementation.
The previous (gorilla mux) based router could deal with regexp in paths, which didn't require such workaround, 
i think it makes sense to make this code `Router` implementation specific

**Review by bmicklea — APPROVED** [review]

**Comment by garagatyi** [comment]
@evoevodin I've commited changes to adress your comments

**Review by voievodin — APPROVED** [review]

**Comment by AndrienkoAleksandr** [comment]
Great job, @garagatyi!

**Review by AndrienkoAleksandr — APPROVED** [review]

**Review by benoitf — APPROVED** [review]

**Pull request merged**